airbyte-server/src/main/java/io/airbyte/server/handlers/SourceHandler.java

```diff
@@ -287,6 +333,109 @@ private ConnectorSpecification getSpecFromSourceDefinitionId(final UUID sourceDe
     return sourceDef.getSpec();
   }
 
+  public SourceDiscoverSchemaRead discoverSchemaForSourceFromSourceId(final SourceDiscoverSchemaRequestBody discoverSchemaRequestBody)
+      throws ConfigNotFoundException, IOException, JsonValidationException {
+    final SourceConnection source = configRepository.getSourceConnection(discoverSchemaRequestBody.getSourceId());
+    final StandardSourceDefinition sourceDef = configRepository.getStandardSourceDefinition(source.getSourceDefinitionId());
+    final String imageName = DockerUtils.getTaggedImageName(sourceDef.getDockerRepository(), sourceDef.getDockerImageTag());
+    final boolean isCustomConnector = sourceDef.getCustom();
+
+    final String configHash = HASH_FUNCTION.hashBytes(Jsons.serialize(source.getConfiguration()).getBytes(
+        Charsets.UTF_8)).toString();
+    final String connectorVersion = sourceDef.getDockerImageTag();
+    final Optional<ActorCatalog> currentCatalog =
+        configRepository.getActorCatalog(discoverSchemaRequestBody.getSourceId(), connectorVersion, configHash);
+    final boolean bustActorCatalogCache = discoverSchemaRequestBody.getDisableCache() != null && discoverSchemaRequestBody.getDisableCache();
+    if (currentCatalog.isEmpty() || bustActorCatalogCache) {
+      final SynchronousResponse<UUID> persistedCatalogId =
+          synchronousSchedulerClient.createDiscoverSchemaJob(
+              source,
+              imageName,
+              connectorVersion,
+              new Version(sourceDef.getProtocolVersion()),
+              isCustomConnector);
+      final SourceDiscoverSchemaRead discoveredSchema = schedulerHandler.retrieveDiscoveredSchema(persistedCatalogId);
```

Comment:
@gosusnp the `retrieveDiscoveredSchema` is also used by `discoverSchemaForSourceFromSourceCreate` in the SchedulerHandler. I didn't move it because wasn't sure if the SourceHandler was the correct place for it. Any ideas to improve here?

Reply:
The intent is to eventually have the content of `schedulerHandler` exploded into specific handlers. With that in mind, I think it makes sense to move `discoverSchemaForSourceFromSourceCreate` to the `SourceHandler` as well.